frontends/pytorch/csrc/c10_dispatch/func_builder.cpp

```diff
@@ -0,0 +1,110 @@
+//===- func_builder.cpp ---------------------------------------------------===//
+//
+// This file is licensed under a pytorch-style license
+// See frontends/pytorch/LICENSE for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "func_builder.h"
+
+#include "mlir-c/StandardAttributes.h"
+#include "mlir-c/StandardTypes.h"
+#include "npcomp-c/Types.h"
+
+using namespace torch_mlir;
+
+MlirType TypeMapper::mapScalarType(c10::ScalarType scalarType) {
+  using c10::ScalarType;
+  switch (scalarType) {
+  case ScalarType::Byte:
+    return mlirIntegerTypeUnsignedGet(context, 8);
+  case ScalarType::Char:
+    return mlirIntegerTypeSignedGet(context, 8);
+  case ScalarType::Short:
+    return mlirIntegerTypeSignedGet(context, 16);
+  case ScalarType::Int:
+    return mlirIntegerTypeSignedGet(context, 32);
+  case ScalarType::Long:
+    return mlirIntegerTypeSignedGet(context, 64);
+  case ScalarType::Bool:
+    return npcompBoolTypeGet(context);
+  case ScalarType::Double:
+    return mlirF64TypeGet(context);
+  case ScalarType::Float:
+    return mlirF32TypeGet(context);
+  case ScalarType::BFloat16:
+    return mlirBF16TypeGet(context);
+  case ScalarType::Half:
+    return mlirF16TypeGet(context);
+  default: {
+    std::stringstream message;
+    message << "unsupported PyTorch scalar type: " << c10::toString(scalarType);
+    throw std::invalid_argument(message.str());
+  }
+  }
+}
+
+MlirType TypeMapper::forwardTensorToType(at::Tensor tensor) {
+  if (!tensor.defined())
```

Comment:
for my own edification, what does "defined" mean in this context?

Reply:
The Tensor can be "null" and the comments specifically say that "care must be taken" for this case. So I took care :)